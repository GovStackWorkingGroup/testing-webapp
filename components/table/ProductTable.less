@import '../../styles/Variables.less';
@import '../../styles/Extends.less';

.table {
  display: flex;
  flex-direction: column;
}

.table-body {
  width: 70vw;
}

.table-results-count {
  height: 52px;
  display: flex;
  align-items: center;
  gap: 5px;
}

.table-results-count > p {
  font-size: 12px;
  color: @text-secondary;
}

.product-table-row-section:extend(.table-row) {
  max-height: 60vh;
  overflow-y: scroll;
  width: 100%;
  display: flex;
  flex-direction: column;
  gap: 1px;
  background-color: @white;
}

.product-table-header:extend(.table-header-text) {
  width: 100%;
  display: grid;
  grid-template-columns: 1fr 4fr 4fr 4fr 4fr 2fr;
  grid-template-rows: 48px;
  border: 1px solid #dfdfea;
  border-radius: 8px 8px 0px 0px;
  align-items: center;
  position: sticky;
}

.product-table-row {
  width: calc(100% + 15px);
  height: 100%;
  display: grid;
  grid-template-columns: 1fr 4fr 4fr 4fr 4fr 2fr;
  border-bottom: 1px solid #dfdfea;
  align-items: center;
}

.product-table-row:hover {
  border-width: 0px 1px 1px 1px;
  border-style: solid;
  border-color: #dfdfea;
  box-shadow: 0px 1px 4px rgba(0, 0, 0, 0.25);
}

.product-table-row-opened {
  background-color: @blue-d-1;
}

.details-arrow {
  display: flex;
  justify-content: flex-end;
  align-items: center;
  font-size: 24px;
  padding-right: 10px;
}

.table-bb-section {
  display: flex;
  align-items: center;
}
.table-bb-section > div:first-of-type:extend(.flex-row) {
  width: 100%;
  align-items: center;
}

.table-bb-image {
  display: flex;
  align-items: center;
  flex-wrap: wrap;
  height: 48px;
  overflow: hidden;
  gap: 10px;
}

.bb-image {
  padding-top: 15px;
  padding-bottom: 15px;
  filter: invert(30%) sepia(96%) saturate(2946%) hue-rotate(221deg)
    brightness(92%) contrast(94%);
}

.overflow-count {
  display: flex;
  align-items: center;
  height: 48px;
  background-color: @blue-10;
  width: 24px;
  height: 24px;
  border-radius: 50%;
}

.overflow-count > p {
  margin: auto;
  color: @accents-and-btn;
}

.product-last-update {
  color: @text-secondary;
}

.data-loader-message {
  text-align: center;
}

.table-error-handling:extend(.table-row) {
  display: flex;
  width: 100%;
  height: @table-row;
  align-items: center;
  text-align: center;
  justify-content: center;
}

.bb-icon-tooltip {
  background-color: @charts-chips-and-perc;
  text-transform: capitalize;
  font-size: 14px;
  padding: 5px 16px;
}
