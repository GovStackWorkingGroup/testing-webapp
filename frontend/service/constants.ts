// Urls
export const COMPLIANCE_TESTING_DETAILS_PAGE =
  'requirements/details/';
export const COMPLIANCE_TESTING_FORM = '/requirements/form';
export const COMPLIANCE_TESTING_RESULT_PAGE = '/requirements';
export const API_TESTING_RESULT_PAGE = '/apiTesting';
export const CONFLUENCE_INSTRUCTIONS_LINK =
  'https://govstack-global.atlassian.net/wiki/spaces/GH/pages/376012801/Instructions+for+Software+Requirements+Compliance';

// Data
export const softwareComplianceFormSteps = [
  {
    label: 'app.software_attributes.label',
    step: 1,
  },
  {
    label: 'table.deployment_compliance.label',
    step: 2,
  },
  {
    label: 'app.api_requirement_specification.label',
    step: 3,
  },
  {
    label: 'app.evaluation_summary.label',
    step: 4,
  },
];

// Local storage
export const SOFTWARE_ATTRIBUTES_STORAGE_NAME = 'software_attributes';
export const DEPLOYMENT_COMPLIANCE_STORAGE_NAME = 'deployment_compliance';
export const INTERFACE_COMPLIANCE_STORAGE_NAME = 'interface_compliance';
export const REQUIREMENT_SPEC_STORAGE_NAME =
  'requirement_specification_compliance';
