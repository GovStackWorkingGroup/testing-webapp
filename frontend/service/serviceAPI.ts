import { FormValuesType } from '../components/form/SoftwareAttributesForm';
import {
  ResultTableSortByType,
  SoftwaresTableSortByType,
} from '../components/table/types';
import {
  BuildingBlockTestSummary,
  ComplianceList,
  ComplianceRequirementsType,
  FormUpdatedObject,
  PATCHSoftwareAttributesType,
  POSTSoftwareAttributesType,
  ProductsListType,
  SoftwareDetailsDataType,
  SoftwareDetailsType,
  SoftwareDraftDetailsType,
  SoftwareDraftToUpdateType,
  SubmitDraftResponseType,
  SubmittingFormResponseType,
} from './types';

export const baseUrl = process.env.API_URL;

type Success<T> = { status: true; data: T };
type Failure = { status: false; error: Error };

const handleFieldsToSort = (
  sortBy: ResultTableSortByType | SoftwaresTableSortByType
) =>
  Object.values(sortBy)
    .filter((order) => order.order !== null)
    .map((sortProperty) => `sort.${sortProperty.field}=${sortProperty.order}`)
    .join('&');

const baseFetch = async (url: string): Promise<ProductsListType> => {
  const response = await fetch(url, {
    method: 'get',
    headers: {
      'Content-Type': 'application/json',
    },
  });

  if (!response.ok) {
    throw new Error(response.statusText);
  }

  return response.json();
};

export const getSoftwaresData = async (
  offset: number,
  sortBy: SoftwaresTableSortByType,
  branch?: string
): Promise<Success<ProductsListType> | Failure> => {
  try {
    const url = new URL(`${baseUrl}/report/`);
    url.searchParams.set('limit', '20');
    url.searchParams.set('offset', offset.toString());
    url.searchParams.set('branch', branch || 'main');

    const sortedParameters = handleFieldsToSort(sortBy);
    for (const param of sortedParameters.split('&')) {
      const [key, value] = param.split('=');
      url.searchParams.set(key, value);
    }

    const actualData = await baseFetch(url.toString());

    return { data: actualData, status: true };
  } catch (error) {
    return {
      status: false,
      error: error as Error,
    };
  }
};

export const getSoftwareListCount = async (
  branch?: string
): Promise<Success<number> | Failure> => {
  try {
    const url = new URL(`${baseUrl}/report/count`);

    // Assuming branch is a query parameter. If not, this can be removed.
    const params = new URLSearchParams({ branch: branch || 'main' });
    url.search = params.toString();

    const response = await fetch(url.toString(), {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (!response.ok) {
      throw new Error(response.statusText);
    }

    const result = await response.json();

    return {
      status: true,
      data: result.count,
    };
  } catch (error) {
    return {
      status: false,
      error: error as Error,
    };
  }
};

export const getBuildingBlockTestResults = async (
  buildingBlockId: string,
  sortBy: ResultTableSortByType
) => {
  const sortedParameters = handleFieldsToSort(sortBy);

  return await fetch(
    `${baseUrl}/report/${buildingBlockId}?${sortedParameters}`,
    {
      method: 'get',
      headers: {
        'Content-Type': 'application/json',
      },
    }
  )
    .then((response) => {
      if (!response.ok) {
        throw new Error(response.statusText);
      }

      return response.json();
    })
    .then<Success<BuildingBlockTestSummary>>((actualData) => {
      return { data: actualData, status: true };
    })
    .catch<Failure>((error) => {
      return { error, status: false };
    });
};

export const getComplianceList = async (offset: number, limit: number) => {

  const accessToken = sessionStorage.getItem('accessToken');

  return await fetch(
    `${baseUrl}/compliance/list?offset=${offset}&limit=${limit}`,
    {
      method: 'get',
      headers: {
        'Content-Type': 'application/json',
        ...(accessToken ? { 'Authorization': `Bearer ${accessToken}` } : {}),
      },
    }
  )
    .then((response) => {
      if (!response.ok) {
        throw new Error(response.statusText);
      }

      return response.json();
    })
    .then<Success<ComplianceList>>((actualData) => {
      return { data: actualData, status: true };
    })
    .catch<Failure>((error) => {
      return { error, status: false };
    });
};

export const getSoftwareDetails = async (softwareName: string) => {
  return await fetch(`${baseUrl}/compliance/${softwareName}/detail`, {
    method: 'get',
    headers: {
      'Content-Type': 'application/json',
    },
  })
    .then((response) => {
      if (!response.ok) {
        throw new Error(response.statusText);
      }

      return response.json();
    })
    .then<Success<SoftwareDetailsType>>((actualData) => {
      return { data: actualData, status: true };
    })
    .catch<Failure>((error) => {
      return { error, status: false };
    });
};

export const saveSoftwareDraft = async (software: FormValuesType) => {
  const formData = new FormData();
  formData.append('softwareName', software.softwareName.value);
  formData.append('logo', software.softwareLogo.value as File);
  formData.append('website', software.softwareWebsite.value);
  formData.append('documentation', software.softwareDocumentation.value);
  formData.append('description', software.toolDescription.value);
  formData.append('email', software.email.value);
  formData.append('version', software.softwareVersion.value);
  formData.append('compliance[0][version]', software.softwareVersion.value);

  return await fetch(`${baseUrl}/compliance/drafts`, {
    method: 'post',
    body: formData,
  })
    .then((response) => {
      if (!response.ok) {
        throw new Error(response.statusText);
      }

      return response.json();
    })
    .then<Success<POSTSoftwareAttributesType>>((response) => {
      return { data: response, status: true };
    })
    .catch<Failure>((error) => {
      return { error, status: false };
    });
};

export const getDraftDetails = async (draftUUID: string) => {
  return await fetch(`${baseUrl}/compliance/drafts/${draftUUID}`, {
    method: 'get',
    headers: {
      'Content-Type': 'application/json',
    },
  })
    .then((response) => {
      if (!response.ok) {
        throw new Error(response.statusText);
      }

      return response.json();
    })
    .then<Success<SoftwareDraftDetailsType>>((actualData) => {
      return { data: actualData, status: true };
    })
    .catch<Failure>((error) => {
      return { error, status: false };
    });
};

export const getComplianceRequirements = async () => {
  return await fetch(`${baseUrl}/compliance/requirements/`, {
    method: 'get',
    headers: {
      'Content-Type': 'application/json',
    },
  })
    .then((response) => {
      if (!response.ok) {
        throw new Error(response.statusText);
      }

      return response.json();
    })
    .then<Success<ComplianceRequirementsType[]>>((actualData) => {
      return { data: actualData, status: true };
    })
    .catch<Failure>((error) => {
      return { error, status: false };
    });
};

export const updateDraftDetailsStepOne = async (
  draftUUID: string,
  data: FormValuesType
) => {
  const formData = new FormData();

  formData.append('softwareName', data.softwareName.value);
  formData.append('logo', data.softwareLogo.value as File);
  formData.append('website', data.softwareWebsite.value);
  formData.append('documentation', data.softwareDocumentation.value);
  formData.append('description', data.toolDescription.value);
  formData.append('email', data.email.value);
  formData.append('compliance[0][version]', data.softwareVersion.value);

  return await fetch(`${baseUrl}/compliance/drafts/${draftUUID}`, {
    method: 'PATCH',
    body: formData,
  })
    .then((response) => {
      if (!response.ok) {
        throw new Error(response.statusText);
      }

      return response.json();
    })
    .then<Success<POSTSoftwareAttributesType>>((actualData) => {
      return { data: actualData, status: true };
    })
    .catch<Failure>((error) => {
      return { error, status: false };
    });
};

export const updateDraftDetailsStepTwo = async (
  draftUUID: string,
  data: SoftwareDraftToUpdateType
) => {
  const formData = new FormData();

  if (
    data.deploymentCompliance?.documentation ||
    data.deploymentCompliance?.documentation === ''
  ) {
    if (data.deploymentCompliance?.documentation instanceof File) {
      formData.append(
        'deploymentCompliance[documentation]',
        data.deploymentCompliance.documentation
      );
    } else {
      formData.append(
        'deploymentCompliance[documentation]',
        data.deploymentCompliance?.documentation
      );
    }
  }

  if (
    data.deploymentCompliance?.deploymentInstructions ||
    data.deploymentCompliance?.deploymentInstructions === ''
  ) {
    if (data.deploymentCompliance?.deploymentInstructions instanceof File) {
      formData.append(
        'deploymentCompliance[deploymentInstructions]',
        data.deploymentCompliance?.deploymentInstructions
      );
    } else {
      formData.append(
        'deploymentCompliance[deploymentInstructions]',
        data.deploymentCompliance?.deploymentInstructions || ''
      );
    }
  }

  return await fetch(`${baseUrl}/compliance/drafts/${draftUUID}`, {
    method: 'PATCH',
    body: formData,
  })
    .then((response) => {
      if (!response.ok) {
        throw new Error(response.statusText);
      }

      return response.json();
    })
    .then<Success<PATCHSoftwareAttributesType>>((actualData) => {
      return { data: actualData, status: true };
    })
    .catch<Failure>((error) => {
      return { error, status: false };
    });
};

export const updateDraftDetailsStepThree = async (
  draftUUID: string,
  data: ComplianceRequirementsType[],
  softwareVersion: string
) => {
  const transformedData = data.map((item) => {
    const {
      bbKey,
      bbName,
      bbVersion,
      dateOfSave,
      requirements,
      interfaceCompliance,
    } = item;
    const test = {
      bbSpecification: bbName,
      bbVersion,
      dateOfSave,
      requirementSpecificationCompliance: {
        crossCuttingRequirements: requirements.crossCutting.map(
          (crossCuttingItem) => ({
            requirement: crossCuttingItem.requirement,
            comment: crossCuttingItem.comment,
            fulfillment: crossCuttingItem.fulfillment,
            _id: crossCuttingItem._id,
          })
        ),
        functionalRequirements: requirements.functional.map(
          (functionalItem) => ({
            requirement: functionalItem.requirement,
            comment: functionalItem.comment,
            fulfillment: functionalItem.fulfillment,
            _id: functionalItem._id,
          })
        ),
      },
      interfaceCompliance: {
        testHarnessResult: interfaceCompliance?.testHarnessResult || '',
        requirements: interfaceCompliance?.requirements || [],
      },
    };

    return {
      [bbKey]: test,
    };
  });

  const transformedDataObject = transformedData.reduce((acc, item) => {
    const bbKey = Object.keys(item)[0];

    if (bbKey) {
      acc[bbKey] = item[bbKey];
    }

    return acc;
  }, {});

  const payload = {
    compliance: {
      version: softwareVersion,
      bbDetails: transformedDataObject,
    },
  };

  return await fetch(`${baseUrl}/compliance/drafts/${draftUUID}`, {
    method: 'PATCH',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(payload),
  })
    .then((response) => {
      if (!response.ok) {
        throw new Error(response.statusText);
      }

      return response.json();
    })
    .then<Success<PATCHSoftwareAttributesType>>((actualData) => {
      return { data: actualData, status: true };
    })
    .catch<Failure>((error) => {
      return { error, status: false };
    });
};

export const submitDraft = async (uniqueId: string) => {
  const payload = { uniqueId };

  return await fetch(`${baseUrl}/compliance/drafts/submit`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(payload),
  })
    .then((response) => {
      if (!response.ok) {
        throw new Error(response.statusText);
      }

      return response.json();
    })
    .then<Success<SubmitDraftResponseType>>((actualData) => {
      return { data: actualData, status: true };
    })
    .catch<Failure>((error) => {
      return { error, status: false };
    });
};

export const getSoftwareDetailsReport = async (id: string) => {
  return await fetch(`${baseUrl}/compliance/forms/${id}`, {
    method: 'get',
    headers: {
      'Content-Type': 'application/json',
    },
  })
    .then((response) => {
      if (!response.ok) {
        throw new Error(response.statusText);
      }

      return response.json();
    })
    .then<Success<SoftwareDetailsDataType>>((actualData) => {
      return { data: actualData, status: true };
    })
    .catch<Failure>((error) => {
      return { error, status: false };
    });
};

export const handleReviewSoftwareForm = async (
  id: string,
  data: FormUpdatedObject,
  type: 'update' | 'accept' | 'reject'
) => {
  const accessToken = sessionStorage.getItem('accessToken');

  return await fetch(`${baseUrl}/compliance/forms/${id}/${type}`, {
    method: 'post',
    body: JSON.stringify(data),
    headers: {
      'Content-Type': 'application/json',
      ...(accessToken ? { Authorization: `Bearer ${accessToken}` } : {}),
    },
  })
    .then((response) => {
      if (!response.ok) {
        throw new Error(response.statusText);
      }

      return response.json();
    })
    .then<Success<SubmittingFormResponseType>>((response) => {
      return { data: response, status: true };
    })
    .catch<Failure>((error) => {
      return { error, status: false };
    });
};

// This endpoint do not connect to project BE
export const checkIfImageUrlExists = async (url: string): Promise<boolean> => {
  try {
    const response = await fetch(url, { method: 'HEAD' });

    return response.status === 200;
  } catch (error) {
    return false;
  }
};

// Fetch file
export const fetchFileDetails = async (file: string) => {
  const filePath = `${baseUrl}/${file}`;
  try {
    const response = await fetch(filePath, {
      method: 'GET',
      mode: 'cors',
      cache: 'no-cache',
    });

    if (!response.ok) {
      throw new Error(`HTTP error! Status: ${response.status}`);
    }

    const contentType = response.headers.get('content-type') || '';
    const blob = await response.blob();

    // Create a File object from the Blob
    const url = new URL(filePath);

    const pathname = url.pathname;

    const parts = pathname.split('/');

    const fileName = parts[parts.length - 1]; // Provide the desired file name
    const file = new File([blob], fileName, { type: contentType });

    return file;
  } catch (error) {
    return undefined;
  }
};
