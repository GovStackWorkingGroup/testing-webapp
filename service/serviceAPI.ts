import {
  ResultTableSortByType,
  SoftwaresTableSortByType,
} from '../components/table/types';
import { BuildingBlockTestSummary, ProductsListType } from './types';

const baseUrl = 'https://api.testing.govstack.global';

type Success<T> = { status: true; data: T };
type Failure = { status: false; error: Error };

const handleFieldsToSort = (
  sortBy: ResultTableSortByType | SoftwaresTableSortByType
) =>
  Object.values(sortBy)
    .filter((order) => order.order !== null)
    .map((sortProperty) => `sort.${sortProperty.field}=${sortProperty.order}`)
    .join('&');

export const getSoftwaresData = async (
  offset: number,
  sortBy: SoftwaresTableSortByType
) => {
  const sortedParameters = handleFieldsToSort(sortBy);

  return await fetch(
    `${baseUrl}/report/?limit=20&offset=${offset}&${sortedParameters}`,
    {
      method: 'get',
      headers: {
        'Content-Type': 'application/json',
      },
    }
  )
    .then((response) => {
      if (!response.ok) {
        throw new Error(response.statusText);
      }
      return response.json();
    })
    .then<Success<ProductsListType>>((actualData) => {
      return { data: actualData, status: true };
    })
    .catch<Failure>((error) => {
      return { error, status: false };
    });
};

export const getSoftwareListCount = async () => {
  return await fetch(`${baseUrl}/report/count`, {
    method: 'get',
    headers: {
      'Content-Type': 'application/json',
    },
  })
    .then(async (response) => {
      if (!response.ok) {
        throw new Error(response.statusText);
      }
      return response.json();
    })
    .then<Success<number>>((actualCount) => {
      return {
        status: true,
        data: actualCount.count,
      };
    })
    .catch<Failure>((error) => {
      return {
        status: false,
        error,
      };
    });
};

export const getBuildingBlockTestResults = async (
  buildingBlockId: string,
  sortBy: ResultTableSortByType
) => {
  const sortedParameters = handleFieldsToSort(sortBy);

  return await fetch(
    `${baseUrl}/report/${buildingBlockId}?${sortedParameters}`,
    {
      method: 'get',
      headers: {
        'Content-Type': 'application/json',
      },
    }
  )
    .then((response) => {
      if (!response.ok) {
        throw new Error(response.statusText);
      }
      return response.json();
    })
    .then<Success<BuildingBlockTestSummary>>((actualData) => {
      return { data: actualData, status: true };
    })
    .catch<Failure>((error) => {
      return { error, status: false };
    });
};
