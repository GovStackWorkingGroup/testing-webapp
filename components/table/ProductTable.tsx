import React, { useCallback, useEffect, useState } from 'react';
import { useIntl } from 'react-intl';
import InfiniteScroll from 'react-infinite-scroll-component';
import { getSoftwaresData, getSoftwareListCount } from '../../service/serviceAPI';
import { ProductsType } from '../../service/types';
import InfiniteScrollCustomLoader from '../InfiniteScrollLoader';
import ProductTableHeader from './ProductTableHeader';
import ProductTableRow from './ProductTableRow';
import TableErrorHandling from './TableErrorHandling';
import { SoftwaresTableSortByType } from './types';

const ProductTable = () => {
  const [productsList, setProductsList] = useState<ProductsType[]>([]);
  const [productListLength, setProductListLength] = useState<number>(0);
  const [isErrorFetchingData, setIsErrorFetchingData] = useState<boolean>(false);
  const [isErrorFetchingCount, setIsErrorFetchingCount] = useState<boolean>(false);
  const [sortBy, setSortBy] = useState<SoftwaresTableSortByType>({
    testApp: { field: 'testApp', order: null },
    lastUpdate: { field: 'lastUpdate', order: null },
    overallCompatibility: { field: 'overallCompatibility', order: null }
  });

  console.log(sortBy);

  const { formatMessage } = useIntl();
  const format = useCallback(
    (id: string) => formatMessage({ id }),
    [formatMessage]
  );

  useEffect(() => {
    fetchData();
  }, [sortBy]);

  const fetchData = async () => {
    const [data, count] = await Promise.all([
      getSoftwaresData(0, sortBy),
      getSoftwareListCount(),
    ]);

    if (data.status) {
      setProductsList(data.data);
    } else {
      setIsErrorFetchingData(true);
    }

    if (count.status) {
      setProductListLength(count.data);
    } else {
      setIsErrorFetchingCount(true);
    }
  };

  const handleLoadMoreData = useCallback(async () => {
    const data = await getSoftwaresData(productsList.length, sortBy);

    if (data.status) {
      setProductsList([...productsList, ...data.data]);
    } else {
      setProductsList([...productsList]);
    }
  }, [productsList]);

  const handleSorting = (tableSortProperties: SoftwaresTableSortByType) => {
    setSortBy(tableSortProperties);
  };

  return (
    <div className='table'>
      <div className='table-results-count'>
        <p>
          {isErrorFetchingCount ? format('table.no_result_count.message') : productListLength}
        </p>
        <p>
          {isErrorFetchingCount ? null : productListLength === 1
            ? format('table.result.label')
            : format('table.result.plural.label')}
        </p>
      </div>
      <div className='table-body'>
        <ProductTableHeader handleSorting={handleSorting}/>
        {isErrorFetchingData ? (
          <TableErrorHandling />
        ) : (
          <div id='scrollableDiv' className='product-table-row-section'>
            <InfiniteScroll
              scrollableTarget='scrollableDiv'
              dataLength={productsList?.length ?? 0}
              next={handleLoadMoreData}
              hasMore={
                isErrorFetchingData
                  ? false
                  : productListLength > productsList?.length
              }
              loader={<InfiniteScrollCustomLoader />}
              style={{ overflowX: 'hidden' }}
            >
              {productsList.map((product, key) => (
                <ProductTableRow product={product} key={`product-${key}`} />
              ))}
            </InfiniteScroll>
          </div>
        )}
      </div>
    </div>
  );
};

export default ProductTable;
