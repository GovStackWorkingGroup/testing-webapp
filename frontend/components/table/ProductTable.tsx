import isEqual from 'lodash/isEqual';
import { useCallback, useEffect, useState } from 'react';
import InfiniteScroll from 'react-infinite-scroll-component';
import { useIntl } from 'react-intl';
import { getSoftwareListCount, getSoftwaresData } from '../../service/serviceAPI';
import { ProductsType } from '../../service/types';
import InfiniteScrollCustomLoader from '../InfiniteScrollLoader';
import ProductTableHeader from './ProductTableHeader';
import ProductTableRow from './ProductTableRow';
import TableErrorHandling from './TableErrorHandling';
import { SoftwaresTableSortByType } from './types';

const defaultSortBy = {
  testApp: { field: 'testApp', order: null },
  lastUpdate: { field: 'lastUpdate', order: null },
  overallCompatibility: { field: 'overallCompatibility', order: null }
} as SoftwaresTableSortByType;

const ProductTable = ({ showAll }: { showAll: boolean }) => {
  const [productsList, setProductsList] = useState<ProductsType[]>([]);
  const [productListLength, setProductListLength] = useState<number>(0);
  const [error, setError] = useState<string | null>(null);
  const [sortBy, setSortBy] = useState<SoftwaresTableSortByType>(defaultSortBy);
  const [loading, setLoading] = useState(true);

  const { formatMessage } = useIntl();
  const format = useCallback((id: string) => formatMessage({ id }), [formatMessage]);

  useEffect(() => {
    const fetchData = async () => {
      try {
        const [data, count] = await Promise.all([
          getSoftwaresData(0, sortBy, 'main', showAll),
          getSoftwareListCount()
        ]);

        if (data.status) {
          setProductsList(data.data);
        }

        if (count.status) {
          setProductListLength(count.data);
        }
      } catch {
        setError('dataFetchFail');
      } finally {
        setLoading(false); // set loading to false once data is fetched or error occurs
      }
    };

    fetchData();
  }, [sortBy]);

  const handleLoadMoreData = useCallback(async () => {
    try {
      const data = await getSoftwaresData(productsList.length, sortBy);

      if (data.status) {
        setProductsList((prev) => [...prev, ...data.data]);
      }
    } catch {
      setError('dataLoadFail');
    }
  }, [productsList, sortBy]);

  const handleSorting = (tableSortProperties: SoftwaresTableSortByType) => {
    if (!isEqual(tableSortProperties, sortBy)) {
      setSortBy(tableSortProperties);
    }
  };

  if (error) {
    return <TableErrorHandling />;
  }

  return (
    <div className='table'>
      <div className='table-results-count'>
        <p>{productListLength}</p>
        <p>{productListLength === 1 ? format('table.result.label') : format('table.result.plural.label')}</p>
      </div>
      <div className='table-body'>
        <ProductTableHeader handleSorting={handleSorting}/>
        <div id='scrollableDiv' className='product-table-row-section'>
          {loading ? (
            <div className="product-table-row-section">
              <InfiniteScrollCustomLoader />
            </div>
          ) : (
            <>
              <InfiniteScroll
                scrollableTarget='scrollableDiv'
                dataLength={productsList.length}
                next={handleLoadMoreData}
                hasMore={productListLength > productsList.length}
                loader={<InfiniteScrollCustomLoader />}
                style={{ overflowX: 'hidden' }}
              >
                {productsList.map((product, index) => (
                  <ProductTableRow product={product} key={index} />
                ))}
              </InfiniteScroll>
            </>
          )}
        </div>
      </div>
    </div>
  );
};

export default ProductTable;
