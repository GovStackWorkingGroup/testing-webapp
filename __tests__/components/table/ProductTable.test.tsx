import { act } from '@testing-library/react';
import ProductTable from '../../../components/table/ProductTable';
import { render } from '../../test-utils/test-utils';
import * as Api from '../../../service/serviceAPI';
import { mockedProduct } from './mockedData/ProductTable';

jest.mock('../../../service/serviceAPI', () => ({
  getData: jest.fn(() => Promise.resolve({ status: true, data: [] })),
  getProductListCount: jest.fn(() =>
    Promise.resolve({ status: true, data: 0 })
  ),
}));

const spyOnGetProductsList = jest.spyOn(Api, 'getData');

describe('Unit tests for ProductTable component:', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render empty placeholder and match snapshot', async () => {
    spyOnGetProductsList.mockImplementationOnce(() =>
      Promise.resolve({ status: true, data: [] })
    );

    const { container } = render(<ProductTable />);

    await act(() => Promise.resolve());

    expect(spyOnGetProductsList).toHaveBeenCalledTimes(1);
    expect(container).toMatchSnapshot();
  });

  it('should render the ProductTable component with data', async () => {
    spyOnGetProductsList.mockImplementationOnce(() =>
      Promise.resolve({ status: true, data: [mockedProduct] })
    );

    const { getByTestId } = render(<ProductTable />);

    await act(() => Promise.resolve());

    expect(spyOnGetProductsList).toHaveBeenCalledTimes(1);
    expect(getByTestId(`product-table-row-${mockedProduct._id.testApp}`)).toBeDefined();
  });
});
