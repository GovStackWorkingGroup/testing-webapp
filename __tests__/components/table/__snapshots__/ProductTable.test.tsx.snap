// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit tests for ProductTable component: should render the ProductTable component and match snapshot 1`] = `
<div>
  <div
    class="table"
  >
    <div
      class="table-results-count"
    >
      <p>
        1
         result
      </p>
    </div>
    <div
      class="table-body"
    >
      <div
        class="product-table-header"
      >
        <div />
        <div>
          <p>
            Product Name
          </p>
        </div>
        <div>
          <p>
            Building Blocks
          </p>
        </div>
        <div>
          <p>
            Last Update
          </p>
        </div>
        <div>
          <p>
            Overall Compatibility
          </p>
        </div>
        <div />
      </div>
      <div
        class="product-table-row-section"
      >
        <div
          class="product-table-row"
        >
          <div
            class="details-arrow"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g>
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M12 13.172l4.95-4.95 1.414 1.414L12 16 5.636 9.636 7.05 8.222z"
                />
              </g>
            </svg>
          </div>
          <div>
            <p>
              Test Product
            </p>
          </div>
          <div
            class="table-bb-section"
          >
            <div>
              <div
                class="table-bb-image"
                style="width: 0px;"
              >
                <img
                  alt="Logo for: Building block label"
                  class="bb-image"
                  data-nimg="1"
                  decoding="async"
                  height="18"
                  loading="lazy"
                  src="/_next/image?url=%2Fimages%2FbuildingBlocks%2FBuilding%20block%20label.png&w=48&q=75"
                  srcset="/_next/image?url=%2Fimages%2FbuildingBlocks%2FBuilding%20block%20label.png&w=32&q=75 1x, /_next/image?url=%2Fimages%2FbuildingBlocks%2FBuilding%20block%20label.png&w=48&q=75 2x"
                  style="color: transparent;"
                  width="18"
                />
              </div>
              <div>
                <div
                  class="overflow-count"
                  data-testid="bb-rest-count"
                >
                  <p>
                    +1
                  </p>
                </div>
              </div>
            </div>
          </div>
          <div
            data-testid="product-last-update"
          >
            <p
              class="product-last-update"
            >
              1/20/1970
            </p>
          </div>
          <div>
            <p />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
